pretrain/dataset.py

```diff
@@ -15,42 +15,156 @@
-    
+    # TODO: Make this class abstract
+    """
     def __init__(
-        self,
-        batch_size=None,
-        sequence_length=None,
-        num_pages=None,
-        tokenizer: AutoTokenizer=None,
+            self,
+            batch_size=None,
+            sequence_length=None,
+            num_pages=None,
+            tokenizer: AutoTokenizer=None,
+            pack_samples: bool=False,
     ):
         self.batch_size = batch_size
         self.sequence_length = sequence_length
         self.num_pages = num_pages
-        self.num_rows_per_page = 100
         self.tokenizer = tokenizer
+        self.pack_samples = pack_samples
+
+        self.num_rows_per_page = 100
+
+        # Buffer to hold pages loaded from the api
+        self.buffer = []
+
+        # Buffer to hold pages already loaded into a batch
+        self.used_buffer = []
+
+        # Buffer to hold padded pages
+        self.padded_buffer = []
+
+    def fetch_data_for_pages(self, pages):
+        """
+        Set the pages to be used to fill the buffer. Then fetch the page data
+        to the buffer.
+        """
+
+        self.pages = pages
 
+        # Empty the buffer if it is not.
         self.buffer = []
 
+        for page in self.pages:
+            self._fetch_data_for_page(page)
+
+
+    def _get_pad_size(self, input_ids):
+        """
+        Get the number of tokens to be padded to the sample to match
+        the max allowed sequence length.
+        If sample packing is activated, then return 1
+        """
+
+        if self.pack_samples:
+            return 1
+
+        sample_size = len(input_ids)
+
+        remainder = (sample_size % self.sequence_length)
+        pad_size = (self.sequence_length - remainder)
+
+        # Apply modulo again to guarantee a pad size of 0 if remainder is 0
+        pad_size = pad_size % self.sequence_length
+
+        return pad_size
+
+    def _refill_padded_buffer(self):
+        """
+        This methods pulls one page from `self.buffer`, pads it and pushs
+        it to the `self.padded_buffer`.
+        """
+
+        while (
+                self.buffer
+                and len(self.padded_buffer) < self.sequence_length
+        ):
+
+            input_ids = []
+
+            # search for EOS token index and cut the buffer at it.
+            EOS_index = self.buffer.index(self.tokenizer.eos_token_id)
+            input_ids = self.buffer[:EOS_index+1]
+            self.buffer =self.buffer[EOS_index+1:]
+
+            self.used_buffer += input_ids
+
+            # Add to padded buffer without the EOS token.
+            self.padded_buffer += input_ids[:-1]
+
+            # Pad
+            self.padded_buffer += [self.tokenizer.eos_token_id] * self._get_pad_size(input_ids=input_ids[:-1])
+
+    def __iter__(self):
+        self.buffer = self.used_buffer + self.buffer
+        self.padded_buffer = []
+
+        # Pad and prepare one page for batching
+        self._refill_padded_buffer()
+
+        return self
```

Comment:
Previously this yielded the stack? This isn't needed anymore?

Reply:
It is not. Actually, the __next__ function is the one that gets called. __iter__() can be used only to reset the iterator if we need to iterate again. This is a standard implementation also for __iter__() returning `self` and `__next__()` handling the yield logic (without using yield)